### ISSUE TITLE
Revise and optimize the reference data implementations

### ISSUE THREAD

**Issue by kaspersorensen** [datacleaner/DataCleaner#182]
Current implementations of Dictionary and SynonymCatalog are quite obscure. There are several unstandard ways of archieving some caching features etc. which I think is both risky and bulky... So let's re-implement these quite simple components using a Guava cache and maybe something like a bloom filter or so. And let's make these features pluggable so that the implementation differences become more composeable and clear.

**Comment by stefanjanssen** [comment]
Personally I feel like bloom filters are way overkill when it comes to performance. As far as I understand, the current filters and transformers take a very painful way to do things, e.g. going through all entries one-by-one to see if it contains a value or not. Let's start by doing something not completely stupid. 

Within Human Inference we already use forked filters and transformers that do have acceptable performance, as well as allowing you to do things like case-insensitive matching. A possible solution would be to port those back to DataCleaner.

**Comment by kaspersorensen** [comment]
I started working on this in the branch ```feature/reference-data-improvements```. So far I did a few things:

 * Added an awareness of start and end of interactions by introducing "openConnection" methods on all the interfaces. This is now more similar to the Datastore API which works a lot better because then we can decide at start time to load stuff into memory or cache or whatever, but also get rid of it when a connection is closed.
 * In quite some cases I have opted for just loading stuff into memory.
 * I have ditched the "FileMonitor" aspect of the implementations. The need for this seems to be alleviated by the connection concept described above.
 * So far I ditched the deserializability of old implementations into new. But on second thought that actually would be pretty important to retain.

I would still like to add:

 * Ability to use any Resource (e.g. HdfsResource or UrlResource) and not just files in the TextFileDictionary and TextFileSynonymCatalog.
 * Deserializability of old objects into new.
